File: go/sudoku/sudoku.go
package sudoku

import (
	"errors"
	"fmt"
	"strings"
	"unicode"
)

var ErrInvalidConfig = errors.New("config must specify 81 cells and contain only 1-9 and \".\"")
var ErrUnsolvable = errors.New("puzzle is not solveable")

// Game represents a standard 9x9 Sudoku game.
type Game [81]int

func LoadGame(config string) (Game, error) {
	var g [81]int
	if len(config) != 81 {
		return g, ErrInvalidConfig
	}

	for i, rn := range config {
		if !unicode.IsDigit(rn) && rn != '.' {
			return g, ErrInvalidConfig
		}
		if rn != '.' {
			g[i] = int(rn - '0')
		} else {
			g[i] = 0
		}
	}
	return g, nil
}

// Solve fills all empty spaces of g or returns a non-nil error if g is unsolvable.
func (g *Game) Solve() error {
	var rows [9][10]bool
	var cols [9][10]bool
	var boxes [9][10]bool
	for i, val := range g {
		r, c := rowColOf(i)
		b := box(r, c)
		rows[r][val] = true
		cols[c][val] = true
		boxes[b][val] = true
	}

	var empties []cell
	for i, val := range g {
		if val != 0 {
			continue
		}

		empty := newCell(i)
		for op := 1; op <= 9; op++ {
			if !rows[empty.row][op] && !cols[empty.col][op] && !boxes[empty.box][op] {
				empty.options[op] = true
			}
		}
		empties = append(empties, empty)
	}

	return backtrackSolve(g, empties)
}

func backtrackSolve(g *Game, empties []cell) error {
	numEmpties := len(empties)
	if numEmpties == 0 {
		return nil
	}

	// Pick the empty cell with the fewest remaining possibilities
	nextEmptyIndex := -1
	nextOptionsCount := -1
	for i, cell := range empties {
		if opCount := cell.optionCount(); nextEmptyIndex == -1 || opCount < nextOptionsCount {
			nextEmptyIndex = i
			nextOptionsCount = opCount
		}
	}

	// We took a wrong turn and need to backtrack
	if nextOptionsCount == 0 {
		return ErrUnsolvable
	}

	// Moves the next candidate cell to the end of empties so we can
	// efficiently pass the remaining empty cells to further calls
	empties[nextEmptyIndex], empties[numEmpties-1] = empties[numEmpties-1], empties[nextEmptyIndex]

	// Try to fill nextCell with each avaialble option
	nextCell := empties[numEmpties-1]
	for value := 1; value <= 9; value++ {
		if !nextCell.options[value] {
			continue
		}

		cellsToRestore := make([]int, 0, len(empties))
		for i := 0; i < len(empties)-1; i++ {
			if nextCell.sharesConstraintWith(empties[i]) {
				cellsToRestore = append(cellsToRestore, i)
				empties[i].options[value] = false
			}
		}

		if backtrackSolve(g, empties[:numEmpties-1]) == nil {
			g[nextCell.index()] = value
			return nil
		}

		for _, index := range cellsToRestore {
			empties[index].options[value] = true
		}
	}

	return ErrUnsolvable
}

func (g Game) IsSolved() bool {
	var rows [9][10]bool
	var cols [9][10]bool
	var boxes [9][10]bool
	for i, val := range g {
		r, c := rowColOf(i)
		b := box(r, c)

		if rows[r][val] {
			return false
		}
		rows[r][val] = true

		if cols[c][val] {
			return false
		}
		cols[c][val] = true

		if boxes[b][val] {
			return false
		}
		boxes[b][val] = true
	}
	return true
}

// String returns a minimal string representation of the board, using "." for empty spaces.
func (g Game) String() string {
	var s strings.Builder
	for _, val := range g {
		if val == 0 {
			s.WriteRune('.')
		} else {
			s.WriteRune(rune('0' + val))
		}
	}
	return s.String()
}

func (g Game) Pformat() string {
	var s strings.Builder
	for i, val := range g {
		if val == 0 {
			s.WriteRune('*')
		} else {
			s.WriteRune(rune('0' + val))
		}

		r, c := rowColOf(i)
		if c != 8 {
			s.WriteRune(' ')
		} else {
			s.WriteRune('\n')
			if r == 2 || r == 5 {
				s.WriteRune('\n')
			}
		}

		if c == 2 || c == 5 {
			s.WriteRune(' ')
		}
	}
	return s.String()
}

func (g Game) Display() {
	fmt.Println(g.Pformat())
}

File: cpp/sudoku/sudoku.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <map>
#include <set>
#include <string>

#include "sudoku/sudoku.h"

template <typename T>
bool set_has(const std::set<T> &s, T val)
{
    return s.find(val) != s.end();
}

class Cell
{
public:
    unsigned short index;
    unsigned short row;
    unsigned short col;
    unsigned short box;
    std::set<unsigned short> options;

    Cell(unsigned short index)
    {
        this->index = index;
        this->row = index / 9;
        this->col = index % 9;
        if (this->row < 3)
        {
            this->box = col / 3;
        }
        else if (this->row < 6)
        {
            this->box = (col / 3) + 3;
        }
        else
        {
            this->box = (col / 3) + 6;
        }
    }

    bool shares_constraint_with(Cell *other)
    {
        return this->row == other->row || this->col == other->col || this->box == other->box;
    }

    void init_options(const std::set<unsigned short> &rows,
                      const std::set<unsigned short> &cols,
                      const std::set<unsigned short> &boxes)
    {
        for (unsigned short i = 1; i <= 9; ++i)
        {
            if (!set_has(rows, i) && !set_has(cols, i) && !set_has(boxes, i))
            {
                this->options.insert(i);
            }
        }
    }

    /** For debugging.. */
    std::string to_string() const
    {
        std::string optionsPart = "";
        if (this->options.empty())
        {
            optionsPart += "EMPTY";
        }
        else
        {
            optionsPart += "[ ";
            for (auto op : this->options)
            {
                optionsPart += std::to_string(op);
                optionsPart += " ";
            }
            optionsPart += "]";
        }

        return "row = " + std::to_string(this->row) + ", col = " + std::to_string(this->col) + ", box = " + std::to_string(this->box) + ", options = " + optionsPart;
    }
};

bool backtrackSolve(sudoku::Sudoku *sudoku,
                    std::vector<Cell *> &empties)
{
    // No more empties, success!
    if (empties.empty())
    {
        return true;
    }

    // Find the cell with the fewest open options
    size_t nextIndex = 0;
    Cell *next = nullptr;
    for (size_t i = 0; i < empties.size(); ++i)
    {
        auto cell = empties[i];
        if (!next || cell->options.size() < next->options.size())
        {
            next = cell;
            nextIndex = i;
        }
    }

    // We've hit a cell with no option, we goofed, time to backtrack
    if (next->options.size() == 0)
    {
        return false;
    }

    // Remove chosen cell
    size_t last = empties.size() - 1;
    std::swap(empties[last], empties[nextIndex]);
    empties.pop_back();

    // Try to fill in the cell with an option and continue
    for (auto option : next->options)
    {
        std::vector<Cell *> affected;
        for (auto cell : empties)
        {
            if (next->shares_constraint_with(cell))
            {
                affected.push_back(cell);
                cell->options.erase(option);
            }
        }

        auto solved = backtrackSolve(sudoku, empties);
        if (solved)
        {
            sudoku->set(next->index, option);
            return true;
        }

        for (auto cell : affected)
        {
            cell->options.insert(option);
        }
    }

    // We we're able to solve past this point, backtrack
    empties.push_back(next);
    return false;
};

sudoku::Sudoku::Sudoku(const std::string &config)
{
    for (auto ch : config)
    {
        if (ch == '.')
        {
            this->cells.push_back(0);
        }
        else
        {
            this->cells.push_back(ch - '0');
        }
    }
}

std::string sudoku::Sudoku::pformat() const
{
    std::string display;
    for (unsigned short i = 0; i < 81; i++)
    {
        unsigned short row = i / 9;
        unsigned short col = i % 9;
        unsigned short val = this->cells[i];

        if (val == 0)
        {
            display.push_back('*');
        }
        else
        {
            display.push_back(val + '0');
        }

        display.push_back(' ');

        if (col == 2 || col == 5)
        {
            display.push_back(' ');
        }

        if (col == 8)
        {
            display.push_back('\n');
            if (row == 2 || row == 5)
            {
                display.push_back('\n');
            }
        }
    }
    return display;
}

std::string sudoku::Sudoku::config() const
{
    std::string config;
    for (auto val : this->cells)
    {
        if (val == 0)
        {
            config.push_back('.');
        }
        else
        {
            config.push_back('0' + val);
        }
    }
    return config;
}

bool sudoku::Sudoku::solve()
{
    std::map<unsigned short, std::set<unsigned short>> rows;
    std::map<unsigned short, std::set<unsigned short>> cols;
    std::map<unsigned short, std::set<unsigned short>> boxes;

    std::vector<Cell *> empties;
    for (size_t index = 0; index < this->cells.size(); ++index)
    {
        auto value = this->cells[index];
        auto cell = new Cell(index);
        if (value != 0)
        {
            rows[cell->row].insert(value);
            cols[cell->col].insert(value);
            boxes[cell->box].insert(value);
            delete cell;
        }
        else
        {
            empties.push_back(cell);
        }
    }

    for (auto cell : empties)
    {
        cell->init_options(rows[cell->row], cols[cell->col], boxes[cell->box]);
    }

    auto solved = backtrackSolve(this, empties);
    for (auto e : empties)
    {
        delete e;
    }
    return solved;
}

void sudoku::Sudoku::set(size_t index, unsigned short option)
{
    this->cells[index] = option;
}